src/table_ops.jl

```diff
@@ -0,0 +1,163 @@
+
+Return the flat index of each row in `table` based on its associated coordinates.
+Dimension columns are determined from the name of each dimension in `dims`.
+It is assumed that the source/destination array has the same dimension order as `dims`.
+
+# Arguments
+- `table`: A table representation of a dimensional array. 
+- `dims`: A `Tuple` of `Dimension` corresponding to the source/destination array.
+- `selector`: The selector type to use for non-numerical/irregular coordinates.
+
+# Example
+```julia
+julia> d = DimArray(rand(256, 256), (X, Y));
+
+julia> t = DimTable(d);
+
+julia> coords_to_index(t, dims(d))
+65536-element Vector{Int64}:
+     1
+     2
+     ⋮
+ 65535
+ 65536
+```
+"""
+function coords_to_index(table, dims::Tuple; selector=DimensionalData.Contains)
+    return _sort_coords(table, dims, selector)
+end
+
+# Find the order of the table's rows according to the coordinate values 
+_sort_coords(table, dims::Tuple, ::Type{T}) where {T <: DimensionalData.Selector} = _sort_coords(_dim_cols(table, dims), dims, T)
+function _sort_coords(coords::NamedTuple, dims::Tuple, ::Type{T}) where {T <: DimensionalData.Selector}
+    ords = _coords_to_ords(coords, dims, T)
+    indices = _ords_to_indices(ords, dims)
+    return indices
+end
+
+# Extract coordinate columns from table
+function _dim_cols(table, dims::Tuple)
+    dim_cols = name(dims)
+    return NamedTuple{dim_cols}(Tables.getcolumn(table, col) for col in dim_cols)
+end
+
+# Extract data columns from table
+function _data_cols(table, dims::Tuple)
+    data_cols = _data_col_names(table, dims)
+    return NamedTuple{Tuple(data_cols)}(Tables.getcolumn(table, col) for col in data_cols)
+end
+
+# Get names of data columns from table
+function _data_col_names(table, dims::Tuple)
+    dim_cols = name(dims)
+    return filter(x -> !(x in dim_cols), Tables.columnnames(table))
+end
+
+# Determine the ordinality of a set of regularly spaced numerical coordinates with a starting locus
+function _coords_to_ords(
+    coords::AbstractVector, 
+    dim::Dimension, 
+    ::Type{<:DimensionalData.Selector},
+    ::Type{<:Real},
+    ::DimensionalData.Start, 
+    ::DimensionalData.Regular)
+    step = (last(dim) - first(dim)) / (length(dim) - 1)
+    return floor.(Int, ((coords .- first(dim)) ./ step) .+ 1)
+end
+
+# Determine the ordinality of a set of regularly spaced numerical coordinates with a central locus
+function _coords_to_ords(
+    coords::AbstractVector, 
+    dim::Dimension, 
+    ::Type{<:DimensionalData.Selector},
+    ::Type{<:Real},
+    ::DimensionalData.Center, 
+    ::DimensionalData.Regular)
+    step = (last(dim) - first(dim)) / (length(dim) - 1)
+    return round.(Int, ((coords .- first(dim)) ./ step) .+ 1)
+end
+
+# Determine the ordinality of a set of regularly spaced numerical coordinates with an end locus
+function _coords_to_ords(
+    coords::AbstractVector, 
+    dim::Dimension, 
+    ::Type{<:DimensionalData.Selector},
```

Comment:
We need dispatch on Contains/Near/At here as At should fail if it's not exact or within atol, and Near should use the end value if the data is outside the lookup range.

Contains falls back to exact At for `Points`.

This is a bit annoying to implement but the behaviour should really be the same as for selectors in other places.

Reply:
I'm not quite sure that I follow. Are you saying that `At` should throw an error when any of the values fall outside of the tolerance? Should this be the case for all methods dispatching on regular numerical dimensions, or just the method you highlighted? 

If so, then we could first compute the floating point index for each value, then compare each value with the integer index produced by `round`/`ceil`/`floor`. If the difference between the two falls outside of the tolerance for any element, then we throw an error.
```julia
float_ords = ((coords .- first(dim)) ./ step) .+ 1
int_ords = ceil.(Int, float_ords)
invalid_value = findfirst(>(tol), abs.(float_ords .- int_ords))
if !isnothing(invalid_value)
    DD.Lookups._selnotfound(dim, rebuild(sel, coords[invalid_value]))
end
return int_ords
```

In the case of `Near`, we could probably just clamp the indices to `[1, length(dim)]` to handle values outside the lookup range.

Reply:
> Are you saying that At should throw an error when any of the values fall outside of the tolerance

Yes, and if a value doesn't exactly match at lookup value when there is no tolerance.

Then, `Near` should never error for any numerical/date etc values (clamping may be an efficient way to do that), and `Contains` should error when none of the intervals in the lookup contain the value (so for `Regular` lookups its `Near` with no clamping).